### DISCUSSION TITLE
Last Closed Won Opp on Account

### DISCUSSION THREAD

**Discussion by tony-jones-tanium** [jamessimone/apex-rollup#527] [Q&A]
Hi Team - We've created a simple Rollup record for calculating the most recent Closed Won Opp on the Account, and populating that ID on the Account. At first we tried this with a Lookup Field on the Account (so the Opp was actually linked, instead of just populating the 18 digit ID), but then also tried populating it to a text field. Neither seem to be working. Any thoughts? 
![Screenshot 2023-10-25 at 11 47 43 AM](https://github.com/jamessimone/apex-rollup/assets/103075625/6a9fc086-5eda-4003-a62e-8ef7a8da6621)

**Comment by jamessimone** [comment]
@tony-jones-tanium I'd have to get a look at a log to see what's going on - one thing I'd _try_ is updating the `Child Object Where Clause` to `StageName = 'Closed Won'`.

**Reply by tony-jones-tanium** [reply]
Tried updating the WHERE clause, but that didnt work. 

Here is the INFO log from the latest run: 
Message
Rollup v1.5.45: RollupDeferredFullRecalcProcessor starting rollup for:
Type: RollupAsyncProcessor.QueueableProcessor
Invocation Point: FROM_FULL_RECALC_LWC
Is Full Recalc: true
Is No Op: false
Is Conductor: false
Inner rollups: 0
Rollup Metadata: {
"attributes" : {
"type" : "Rollup__mdt",
"url" : "/services/data/v59.0/sobjects/Rollup__mdt/m0k59000000CbWIAA0"
},
"SplitConcatDelimiterOnCalcItem__c" : false,
"CalcItem__c" : "Opportunity",
"LookupObject__c" : "Account",
"RollupToUltimateParent__c" : false,
"LookupFieldOnCalcItem__c" : "AccountId",
"MasterLabel" : "Latest Closed Won Oppty",
"CalcItemWhereClause__c" : "stagename = 'CLOSED WON'",
"RollupOrderBys__r" : {
"totalSize" : 1,
"done" : true,
"records" : [ {
"attributes" : {
"type" : "RollupOrderBy__mdt",
"url" : "/services/data/v59.0/sobjects/RollupOrderBy__mdt/m0h59000000CbJdAAK"
},
"Rollup__c" : "m0k59000000CbWIAA0",
"FieldName__c" : "CloseDate",
"Ranking__c" : 0.0,
"SortOrder__c" : "Descending",
"Id" : "m0h59000000CbJdAAK",
"NullSortOrder__c" : "NULLS FIRST"
} ]
},
"IsFullRecordSet__c" : true,
"ShouldRunWithoutCustomSettingEnabled__c" : false,
"RollupOperation__c" : "FIRST",
"RollupFieldOnLookupObject__c" : "Latest_Closed_Won_Oppty__c",
"Id" : "m0k59000000CbWIAA0",
"RollupFieldOnCalcItem__c" : "Opportunity_18ID__c",
"RollupControl__c" : "m0g7V000000fxTYQAY",
"DeveloperName" : "Latest_Closed_Won_Oppty",
"IsRollupStartedFromParent__c" : false,
"LookupFieldOnLookupObject__c" : "Id"
}
Rollup Control: {
"attributes" : {
"type" : "RollupControl__mdt",
"url" : "/services/data/v59.0/sobjects/RollupControl__mdt/m0g7V000000fxTY"
},
"OnlyRunInFlowContexts__c" : null,
"MaxRollupRetries__c" : 20,
"ShouldRunAs__c" : "Queueable",
"Label" : "Opportunity Rollup Control",
"MasterLabel" : "Opportunity Rollup Control",
"ShouldAbortRun__c" : false,
"Language" : "en_US",
"MaxLookupRowsBeforeBatching__c" : 500,
"DeveloperName" : "Opportunity_Rollup_Controll",
"BatchChunkSize__c" : 200,
"NamespacePrefix" : null,
"MaxNumberOfQueries__c" : 50,
"ShouldDuplicateRulesBeIgnored__c" : false,
"ReplaceCalcItemsAsyncWhenOverCount__c" : 1,
"ShouldRunSingleRecordsSynchronously__c" : false,
"IsDevEdOrTrialOrg__c" : false,
"TriggerOrInvocableName__c" : null,
"MaxParentRowsUpdatedAtOnce__c" : 500,
"SystemModstamp" : "2023-02-06T03:05:36.000+0000",
"IsRollupLoggingEnabled__c" : true,
"ShouldSkipResettingParentFields__c" : true,
"QualifiedApiName" : "Opportunity_Rollup_Controll",
"Id" : "m0g7V000000fxTY",
"IsMergeReparentingEnabled__c" : true,
"MaxQueryRows__c" : 8000
}
Calc Items: "64 items"
Full Recalc Processor: "RollupDeferredFullRecalcProcessor"
Is Processed?: true

DEBUG Log: 
Message
Rollup v1.5.45: gathering additional calc items with query:
SELECT opportunity_18id__c,accountid,stagename,closedate,CurrencyIsoCode
FROM Opportunity
WHERE Id != :objIds
AND AccountId = :recordIds

**Reply by jamessimone** [reply]
@tony-jones-tanium I'd refresh the page - you can see it's still using the old where clause. Also, this is only the "start" of the log - there should be a bunch of other Batch Apex Chunks that get generated since you're doing a full recalc

**Reply by tony-jones-tanium** [reply]
@jamessimone Was able to get it to run on a full recalc, however it only updated 1 record, when it should have updated 50+: 
![image](https://github.com/jamessimone/apex-rollup/assets/103075625/6641bfed-e9d3-49ee-8900-782884f8a3cd)

**Reply by jamessimone** [reply]
@tony-jones-tanium let me know how I can help. I can't do much with screenshots, though, aside from saying that there's isn't anything obviously wrong with your setup.

**Reply by tony-jones-tanium** [reply]
Does is help having an excel file of the Logs? I have attached it here. 

Just seems odd to only update a single record - so we know its partially working. 
[report1698269069552.csv](https://github.com/jamessimone/apex-rollup/files/13170799/report1698269069552.csv)

**Reply by jamessimone** [reply]
@tony-jones-tanium have you tried updating the where clause? It's still showing `stagename = 'CLOSED WON'` and it may be that the field name not conforming to the API name casing for `StageName` is causing this issue.

**Reply by tony-jones-tanium** [reply]
Yeah, thats been updated, and still only the single record has been updated: 
<img width="1213" alt="image" src="https://github.com/jamessimone/apex-rollup/assets/103075625/baa5bada-470e-4c4f-bba9-0a7379851b62">

**Reply by tony-jones-tanium** [reply]
I also tried updating the package. Since then, the number of Log Entries has greatly increased, however, its still not updating the Parent Object as it should. Attaching a report of the Error Logs to see if thats helpful. 
<img width="1331" alt="image" src="https://github.com/jamessimone/apex-rollup/assets/103075625/7a8e85e6-f82a-4c56-9432-bc7d72c4ac86">
[report1698337012707.csv](https://github.com/jamessimone/apex-rollup/files/13180063/report1698337012707.csv)

**Reply by jamessimone** [reply]
I'm seeing an optimization here which will help, but I'm not sure it answers the underlying question of why you're only seeing one update get made. I'll include the optimization with the next version. In my own testing of full recalcs, I have the exact same rollup you have setup, and it's correctly rolling up for all parents.

**Comment by tony-jones-tanium** [comment]
Would there be any issues with the Rollup Operation being "FIRST"? Or anything to do with Rollup Order Bys: 
<img width="1175" alt="image" src="https://github.com/jamessimone/apex-rollup/assets/103075625/6232a4e3-471c-4aba-b366-210eb1732a7e">

**Reply by jamessimone** [reply]
I don't think so. As I said, I replicated this exact rollup for testing and didn't have any issues. I think part of the problem is the time out that's occurring, and I've pushed up a [fix for this](https://github.com/jamessimone/apex-rollup/commit/b3abe93a699841af5380a946c2e9d2955bfe1975) that will come out with the next version.
